### PR TITLE
[FLINK-38568] [mysql-cdc] [cdc-base] Optimize binlog split lookup using binary search

### PR DESCRIPTION
**Pull request by huyuanfeng2018**
## What is the purpose of the change

Optimize CDC binlog split lookup from O(n) to O(log n) using binary search.

## Brief change log

- Add `sortFinishedSplitInfos()` and `findSplitByKeyBinary()` methods
- Update `BinlogSplitReader` to use binary search instead of linear search
- Add comprehensive unit tests

## Performance Impact

- **Time Complexity**: O(n) → O(log n)

## Verifying this change

Added 6 new unit tests covering various scenarios including edge cases and consistency verification with linear search.

### CODE REVIEW

**Comment by huyuanfeng2018** [comment]
hi , @ruanhang1993. Could you have the time to review this Pr? Thank you very much ~

**Comment by loserwang1024** [comment]
@huyuanfeng2018 please also add this improvement to base framework

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/utils/RecordUtils.java [hunk 1]
```diff
@@ -454,39 +453,153 @@ public static BinlogOffset getBinlogPosition(Map<String, ?> offset) {
...
+     * @param key The chunk key to search for
+     * @return The split containing the key, or null if not found
+     */
+    public static FinishedSnapshotSplitInfo findSplitByKeyBinary(
+            List<FinishedSnapshotSplitInfo> sortedSplits, Object[] key) {
+
+        if (sortedSplits == null || sortedSplits.isEmpty()) {
+            return null;
+        }
+
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/utils/RecordUtils.java]: Current PR is already in a good shape, I only have two possible suggestions:

(1) Considering that most tables use auto-increment primary keys and that INSERT operations outnumber UPDATE operations in the binlog, the majority of events in the binlog will typically correspond to either the last split or the first split. Leveraging this data locality, we can further optimize by first checking whether the changelog matches the first or last split and then performing binary search on the remaining splits. 

(2) We can also do same optimization for IncrementalSourceStreamFetcher which is used by other cdc connectors 

WDYT? @huyuanfeng2018

**Comment by huyuanfeng2018** [comment]
Thanks for @loserwang1024  and @leonardBang  for the review. I totally agree with this optimization for the auto-increment scenario. In this case, almost only one comparison is needed, which is very cool. At the same time, I have added this optimization in other sources as well, except for MongoDB and Oracle.  Please take the time to review it again~

**Comment by huyuanfeng2018** [comment]
CI failed. Apache Download CDN has already removed the Flink 1.20.1 binaries pkg. Perhaps we need to upgrade the dependency version.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/utils/SourceRecordUtils.java [hunk 2]
```diff
@@ -206,4 +175,165 @@ public static HistoryRecord getHistoryRecord(SourceRecord schemaRecord) throws I
...
+     *
+     * <p>To leverage data locality for append-heavy workloads (e.g. auto-increment PKs), this
+     * method checks the first and last splits before applying binary search to the remaining
+     * subset.
+     *
+     * @param sortedSplits List of splits sorted by splitStart (MUST be sorted!)
+     * @param key The chunk key to search for
+     * @return The split containing the key, or null if not found
+     */
+    public static FinishedSnapshotSplitInfo findSplitByKeyBinary(
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/utils/SourceRecordUtils.java]: Thanks @huyuanfeng2018 for the update, the updating looks good, one minor comment: this Utils class is a little long after this PR, could we extract a new Utils like SplitKeyUtils to make the code more readable?

Comment (huyuanfeng2018) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/utils/SourceRecordUtils.java]: Thanks for the quick review~

done.

**Comment by leonardBang** [comment]
> CI failed. Apache Download CDN has already removed the Flink 1.20.1 binaries pkg. Perhaps we need to upgrade the dependency version.

@huyuanfeng2018 Could you append a commit to  bump flink version to 1.20.3 to fix this issue ? I've thecked that https://dlcdn.apache.org/flink/flink-1.20.3/ should be okay.

**Comment by huyuanfeng2018** [comment]
> > CI failed. Apache Download CDN has already removed the Flink 1.20.1 binaries pkg. Perhaps we need to upgrade the dependency version.
> 
> @huyuanfeng2018 Could you append a commit to bump flink version to 1.20.3 to fix this issue ? I've thecked that https://dlcdn.apache.org/flink/flink-1.20.3/ should be okay.

ok

**Comment by leonardBang** [comment]
@huyuanfeng2018 I like your community cooperation style, now we can rebase this PR to latest master and convert to normal one

**Comment by huyuanfeng2018** [comment]
> @huyuanfeng2018 I like your community cooperation style, now we can rebase this PR to latest master and convert to normal one

Already rebased onto master. Waiting for CI to finish, I will change the PR status from Draft to Ready.
